osutil/resolve_path.go

```diff
@@ -0,0 +1,117 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func dumbJoin(a, b string) string {
+	if strings.HasSuffix(a, "/") {
+		return a + b
+	} else {
+		return a + "/" + b
+	}
+}
+
+func resolvePathInSysrootRec(sysroot, path string, validInAnySysroot bool, symlinkRecursion int) (string, error) {
+	if path == "" || path == "/" {
+		// Relative paths are taken from sysroot
+		return "/", nil
+	}
+
+	if strings.HasSuffix(path, "/") {
+		path = path[:len(path)-1]
+	}
+
+	dir, file := filepath.Split(path)
+	resolvedDir, err := resolvePathInSysrootRec(sysroot, dir, validInAnySysroot, symlinkRecursion)
+	if err != nil {
+		return "", err
+	}
+	if file == "" {
+		return resolvedDir, nil
+	}
+	if file == "." {
+		return resolvedDir, nil
+	}
+	if file == ".." {
+		if validInAnySysroot && (resolvedDir == "/") {
+			return "", fmt.Errorf("invalid escaping path")
+		}
+		upperDir, _ := filepath.Split(resolvedDir)
+		return upperDir, nil
+	}
+
+	fileInResolvedDir := dumbJoin(resolvedDir, file)
+
+	realPath := dumbJoin(sysroot, fileInResolvedDir)
+	st, err := os.Lstat(realPath)
+	if err != nil {
+		return "", err
+	}
+
+	if st.Mode()&os.ModeSymlink != 0 {
+		if symlinkRecursion < 0 {
+			return "", fmt.Errorf("Maximum recursion reached when reading symlinks")
+		}
+		target, err := os.Readlink(realPath)
+		if err != nil {
+			return "", err
+		}
+		if filepath.IsAbs(target) {
+			if validInAnySysroot {
+				return "", fmt.Errorf("Invalid absolute symlink")
+			}
+			return resolvePathInSysrootRec(sysroot, target, validInAnySysroot, symlinkRecursion-1)
+		} else {
+			return resolvePathInSysrootRec(sysroot, dumbJoin(resolvedDir, target), validInAnySysroot, symlinkRecursion-1)
+		}
+	}
+
+	return fileInResolvedDir, nil
+}
+
+// Resolve a path within a sysroot
+//
+// In a sysroot, abolute symlinks should be relative to the sysroot
+// rather than `/`. Also paths with multiple `..` that would escape
+// the sysroot should not do so.
+//
+// The return path is the path within the sysroot. filepath.Join() has
+// to be used to get the path in the sysroot.
+func ResolvePathInSysroot(sysroot, path string) (string, error) {
+	return resolvePathInSysrootRec(sysroot, path, false, 255)
+}
+
+// Resolve a path within a pseudo sysroot
+//
+// Like ResolvePathInSysroot(), it resolves path as if it was a sysroot.
+// However, any escaping relative path, or absolute symlink generates
+// an error.
+//
+// This is useful when a path is expected to be relative only and sees
+// any "attempt" to escape the sysroot as a malformed path.
+func ResolvePathInSysrootValidAnywhere(sysroot, path string) (string, error) {
```

Comment:
I'm still unsure what ValidAnywhere means here? would EscapesAreInvalid be clearer?

Reply:
It was to mean it has the same meaning whether it is seen as a sysroot or a normal directory.

I have renamed it `ResolvePathNoEscape`. I think it is more obvious like that.